src/x86.rs

```diff
@@ -132,222 +132,228 @@ impl<V: Vector> VectorHash<V> {
-
-                while eq != 0 {
-                    let chunk = chunk.add(eq.trailing_zeros() as usize);
-                    if $memcmp(chunk, needle, self.size() - 1) {
-                        return true;
-                    }
-
-                    eq = bits::clear_leftmost_set(eq);
-                }
-
-                false
-            }
+            hash.pop(*unsafe { haystack.get_unchecked(start) });
+        }
 
-            #[inline]
-            #[target_feature(enable = "avx2")]
-            unsafe fn vector_search_in<V: Vector>(
-                &self,
-                haystack: &[u8],
-                hash: &VectorHash<V>,
-                next: unsafe fn(&Self, &[u8]) -> bool,
-            ) -> bool {
-                debug_assert!(haystack.len() >= self.size());
-
-                let lanes = mem::size_of::<V>();
-                let end = haystack.len() - self.size() + 1;
-
-                if end < lanes {
-                    return next(self, haystack);
-                }
-
-                let mut chunks = haystack[..end].chunks_exact(lanes);
-                while let Some(chunk) = chunks.next() {
-                    if self.vector_search_in_chunk(haystack, hash, chunk.as_ptr(), -1) {
-                        return true;
-                    }
-                }
-
-                let remainder = chunks.remainder().len();
-                if remainder > 0 {
-                    let start = haystack.as_ptr().add(end - lanes);
-                    let mask = -1 << (lanes - remainder);
-
-                    if self.vector_search_in_chunk(haystack, hash, start, mask) {
-                        return true;
-                    }
-                }
-
-                false
-            }
+        false
+    }
 
-            #[inline]
-            #[target_feature(enable = "avx2")]
-            unsafe fn sse2_search_in(&self, haystack: &[u8]) -> bool {
-                self.vector_search_in(haystack, &self.sse2_hash, Self::scalar_search_in)
+    #[inline]
+    #[target_feature(enable = "avx2")]
+    unsafe fn vector_search_in_chunk<V: Vector>(
+        &self,
+        haystack: &[u8],
+        hash: &VectorHash<V>,
+        start: *const u8,
+        mask: i32,
+    ) -> bool {
+        let first = Vector::loadu_si(start.cast());
+        let last = Vector::loadu_si(start.add(self.position).cast());
+
+        let eq_first = Vector::cmpeq_epi8(hash.first, first);
+        let eq_last = Vector::cmpeq_epi8(hash.last, last);
+
+        let eq = Vector::and_si(eq_first, eq_last);
+        let mut eq = (Vector::movemask_epi8(eq) & mask) as u32;
+
+        let start = start as usize - haystack.as_ptr() as usize;
+        let chunk = haystack.as_ptr().add(start + 1);
+        let needle = self.needle.as_ref().as_ptr().add(1);
+
+        while eq != 0 {
+            let chunk = chunk.add(eq.trailing_zeros() as usize);
+            let equal = match N::SIZE {
+                Some(0) => unreachable!(),
```

Comment:
We may want to decide more formally on the minimum length for a needle. I think we may want this to actually be 2 rather than 1. Perhaps an issue for later though.

Reply:
Yeah I only gated `0` here because I added an early exit check earlier.

Reply:
👍